File: codeflix/Catalog/src/Catalog.Domain/Entity/Category.cs
﻿using Catalog.Domain.Exceptions;
using Catalog.Domain.SeedWork;
using Catalog.Domain.Validation;
using static System.String;

namespace Catalog.Domain.Entity;

public class Category : AggregateRoot
{
    public string Name { get; private set; }
    public string Description { get; private set; }
    public DateTime CreatedAt { get; private set; }
    public bool IsActive { get; private set; }


    public Category(string name, string description, bool isActive = true) : base()
    {
        Name = name;
        Description = description;
        CreatedAt = DateTime.Now;
        IsActive = isActive;
        Validate();
    }

    public void Update(string name, string? description = null)
    {
        Name = name;
        Description = description ?? Description;
        Validate();
    }

    public void Activate()
    {
        IsActive = true;
        Validate();
    }

    public void Deactivate()
    {
        IsActive = false;
        Validate();
    }

    private void Validate()
    {
        DomainValidation.NotNullOrEmpty(Name, nameof(Name));
        DomainValidation.MinLength(Name, 3, nameof(Name));
        DomainValidation.MaxLength(Name, 255, nameof(Name));
        DomainValidation.NotNull(Description, nameof(Description));
        DomainValidation.MaxLength(Description, 10_000, nameof(Description));
    }
}

File: codeflix/catalog_rust/src/catalog_domain/entity/category.rs
use chrono::{DateTime, Utc};
use uuid::Uuid;

use crate::catalog_domain::EntityValidationException;

#[derive(Debug, serde::Serialize, serde::Deserialize)]
pub struct Category {
    pub id: Uuid,
    pub name: String,
    pub description: String,
    pub is_active: bool,
    pub created_at: DateTime<Utc>,
}

impl Category {
    pub fn new(
        name: String,
        description: String,
        is_active: Option<bool>,
    ) -> Result<Self, EntityValidationException> {
        let category = Self {
            id: Uuid::new_v4(),
            name,
            description,
            is_active: is_active.unwrap_or(true),
            created_at: Utc::now(),
        };

        validate(&category)?;
        Ok(category)
    }

    pub fn update(
        &mut self,
        name: String,
        description: Option<String>,
    ) -> Result<(), EntityValidationException> {
        self.name = name;
        if let Some(desc) = description {
            self.description = desc;
        }
        validate(self)
    }

    pub fn activate(&mut self) -> Result<(), EntityValidationException> {
        self.is_active = true;
        validate(self)
    }

    pub fn deactivate(&mut self) -> Result<(), EntityValidationException> {
        self.is_active = false;
        validate(self)
    }
}

fn validate(category: &Category) -> Result<(), EntityValidationException> {
    use crate::catalog_domain::validation::domain_validation::DomainValidation;

    DomainValidation::not_empty_or_whitespace(&category.name, "Name")?;
    DomainValidation::min_length(&category.name, "Name", 3)?;
    DomainValidation::max_length(&category.name, "Name", 255)?;
    DomainValidation::max_length(&category.description, "Description", 10000)?;

    Ok(())
}
